### ISSUE TITLE
Powerwall battery level % and battery remaining (Wh) randomly jump to zero

### ISSUE THREAD

**Issue by jaymunro** [alandtse/tesla#475]
## Version of the custom_component
v3.9.1

## Configuration
not sure what to put here. configuration.yaml ?

## Describe the bug
The Powerwall battery level % and Wh remaining jump to zero for one 10s polling cycle about 10-30 times each hour. This kinda buggers up automation triggers. This has been occurring since removing the previous version (as per recommendation) and installing the current one yesterday.

<img width="537" alt="Screenshot 2023-01-06 at 4 53 11 PM" src="https://user-images.githubusercontent.com/1166024/210926717-0989f327-c93a-4779-a826-95067f0ac613.png">

The debug log below should contain two such cases: 16:37:26 & 16:37:47 as shown in this history graph:
<img width="1191" alt="Screenshot 2023-01-06 at 4 42 28 PM" src="https://user-images.githubusercontent.com/1166024/210925705-03f1bf35-471f-412d-8d09-864736ead9c8.png">

## Debug log
[home-assistant_tesla_custom_2023-01-06T03-41-30.439Z.log](https://github.com/alandtse/tesla/files/10357162/home-assistant_tesla_custom_2023-01-06T03-41-30.439Z.log)

**Comment by jaymunro** [comment]
Examining the log, I see the entry at 
`2023-01-06 16:37:14.040 DEBUG (MainThread) [teslajsonpy.connection] 200: {"response":{"site_name":"My Home","energy_left":13893,"total_pack_energy":13893,"grid_status":"Active","backup":{"backup_reserve_percent":16,"events":[{"timestamp":"2022-12-23T17:43:50+13:00","duration":77979},{"timestamp":"2022-10-04T16:30:26+13:00","duration":212798},{"timestamp":"2022-07-17T00:41:57+12:00","duration":20056},{"timestamp":"2022-07-15T12:48:29+12:00","duration":638086},{"timestamp":"2022-07-14T16:16:05+12:00","duration":111709},{"timestamp":"2022-07-13T12:30:16+12:00","duration":31192},{"timestamp":"2022-07-02T17:21:52+12:00","duration":92101},{"timestamp":"2022-07-02T16:37:22+12:00","duration":133100},{"timestamp":"2022-07-02T16:34:26+12:00","duration":105956},{"timestamp":"2022-07-02T13:38:34+12:00","duration":153801},{"timestamp":"2022-07-02T13:36:25+12:00","duration":21030}],"events_count":0,"total_events":0},"user_settings":{"storm_mode_enabled":true,"powerwall_onboarding_settings_set":true,"powerwall_tesla_electric_interested_in":false,"sync_grid_alert_enabled":false,"breaker_alert_enabled":false},"components":{"solar":true,"solar_type":"pv_panel","battery":true,"grid":true,"backup":true,"gateway":"teg","load_meter":true,"tou_capable":true,"storm_mode_capable":true,"flex_energy_request_capable":false,"car_charging_data_supported":false,"off_grid_vehicle_charging_reserve_supported":false,"vehicle_charging_performance_view_enabled":false,"vehicle_charging_solar_offset_view_enabled":false,"battery_solar_offset_view_enabled":true,"solar_value_enabled":true,"energy_value_header":"Energy Value","energy_value_subheader":"Estimated Value","show_grid_import_battery_source_cards":true,"backup_time_remaining_enabled":true,"battery_type":"ac_powerwall","configurable":false,"grid_services_enabled":false,"customer_preferred_export_rule":"pv_only"},"default_real_mode":"autonomous","operation":"autonomous","installation_date":"2022-07-01T15:15:19+12:00","power_reading":[{"timestamp":"2023-01-06T16:19:51+13:00","load_power":2070,"solar_power":2485,"grid_power":-315,"battery_power":-100,"generator_power":0}],"battery_count":1}}
`

Next is:
`2023-01-06 16:37:25.795 DEBUG (MainThread) [teslajsonpy.connection] 200: {"response":{"site_name":"My Home","energy_left":0,"total_pack_energy":1,"grid_status":"Active","backup":{"backup_reserve_percent":16,"events":[{"timestamp":"2022-12-23T17:43:50+13:00","duration":77979},{"timestamp":"2022-10-04T16:30:26+13:00","duration":212798},{"timestamp":"2022-07-17T00:41:57+12:00","duration":20056},{"timestamp":"2022-07-15T12:48:29+12:00","duration":638086},{"timestamp":"2022-07-14T16:16:05+12:00","duration":111709},{"timestamp":"2022-07-13T12:30:16+12:00","duration":31192},{"timestamp":"2022-07-02T17:21:52+12:00","duration":92101},{"timestamp":"2022-07-02T16:37:22+12:00","duration":133100},{"timestamp":"2022-07-02T16:34:26+12:00","duration":105956},{"timestamp":"2022-07-02T13:38:34+12:00","duration":153801},{"timestamp":"2022-07-02T13:36:25+12:00","duration":21030}],"events_count":0,"total_events":0},"user_settings":{"storm_mode_enabled":true,"powerwall_onboarding_settings_set":true,"powerwall_tesla_electric_interested_in":false,"sync_grid_alert_enabled":false,"breaker_alert_enabled":false},"components":{"solar":true,"solar_type":"pv_panel","battery":true,"grid":true,"backup":true,"gateway":"teg","load_meter":true,"tou_capable":true,"storm_mode_capable":true,"flex_energy_request_capable":false,"car_charging_data_supported":false,"off_grid_vehicle_charging_reserve_supported":false,"vehicle_charging_performance_view_enabled":false,"vehicle_charging_solar_offset_view_enabled":false,"battery_solar_offset_view_enabled":true,"solar_value_enabled":true,"energy_value_header":"Energy Value","energy_value_subheader":"Estimated Value","show_grid_import_battery_source_cards":true,"backup_time_remaining_enabled":true,"battery_type":"ac_powerwall","configurable":false,"grid_services_enabled":false,"customer_preferred_export_rule":"pv_only"},"default_real_mode":"autonomous","operation":"autonomous","installation_date":"2022-07-01T15:15:19+12:00","power_reading":[{"timestamp":"2023-01-06T16:35:20+13:00","load_power":2084,"solar_power":2525,"grid_power":-341,"battery_power":-100,"generator_power":0}],"battery_count":0}}
`

Note the following in the second reading above:
`"energy_left":0,
"total_pack_energy":1,
"battery_count":0
`
Other response values seem valid.

There is obviously something wrong in the API call or response. Until the cause is found, perhaps a quick workaround would be to ignore all responses with a battery_count of zero?

**Comment by jaymunro** [comment]
On a hunch I tried changing the MIN_SCAN_INTERVAL (restarted & confirmed interval change in the UI) from 10s to 30s in case there was something unexpected on the tesla API with rates. Frequency of false returns decreased but were still present so no luck there.

I also tried disabling the official Tesla integration (www.home-assistant.io/integrations/powerwall) in case of conflicts, but the error returns continued.

**Comment by stevesreed** [comment]
I've been having this issue for a while. I still have the official integration running and it does not seem to have the 0 readings.

**Comment by jaymunro** [comment]
I quite agree. I have the official Tesla Powerwall app running too without the zero values. It uses a different auth mechanism though and has much less functionality

**Comment by jaymunro** [comment]
After formatting the logs I noticed something potentially very odd in the reading timestamps

```
2023-01-06 16:37:14.040 DEBUG (MainThread) [teslajsonpy.connection] 200:
{
  "response":{
    "site_name":"My Home",
    "energy_left":13893,
    "total_pack_energy":13893,
    "grid_status":"Active",
...
    "default_real_mode":"autonomous",
    "operation":"autonomous",
    "installation_date":"2022-07-01T15:15:19+12:00",
    "power_reading":[{
      "timestamp":"2023-01-06T16:19:51+13:00",
      "load_power":2070,
      "solar_power":2485,
      "grid_power":-315,
      "battery_power":-100,
      "generator_power":0
    }],
    "battery_count":1
  }
}
```

```
2023-01-06 16:37:25.795 DEBUG (MainThread) [teslajsonpy.connection] 200:
{
  "response":{
    "site_name":"My Home",
    "energy_left":0,
    "total_pack_energy":1,
    "grid_status":"Active",
...
    "default_real_mode":"autonomous",
    "operation":"autonomous",
    "installation_date":"2022-07-01T15:15:19+12:00",
    "power_reading":[{
      "timestamp":"2023-01-06T16:35:20+13:00",
      "load_power":2084,
      "solar_power":2525,
      "grid_power":-341,
      "battery_power":-100,
      "generator_power":0
    }],
    "battery_count":0
  }
}
 ```

Note how the DEBUG timestamps differ greatly from the `power_reading` timestamps:
```
2023-01-06 16:37:14.040 DEBUG
2023-01-06T16:19:51+13:00
```
```
2023-01-06 16:37:25.795 DEBUG
2023-01-06T16:35:20+13:00
```

**Comment by jaymunro** [comment]
Parsing the log file I see `power_reading` timestamps seem to wander randomly all over the place. Does this suggest that there is something missing in the API call to retrieve the correct reading?
```
2023-01-06 16:37:14.040 DEBUG: 2023-01-06T16:19:51+13:00
2023-01-06 16:37:25.795 DEBUG: 2023-01-06T16:35:20+13:00
2023-01-06 16:37:37.683 DEBUG: 2023-01-06T16:12:07+13:00
2023-01-06 16:37:47.899 DEBUG: 2023-01-06T16:35:27+13:00
2023-01-06 16:37:57.878 DEBUG: 2023-01-06T16:24:51+13:00
etc.
```

**Comment by monkga1** [comment]
I also have the same issue.

**Comment by purcell-lab** [comment]
This is a duplicate of #296 , I have posted some workaround templates there.

This issue is in the API reporting from upstream.

Here is my workaround templates:
```yaml
- platform: template
  sensors:
    filtered_powerwall_soc:
      value_template: >
        {% if states('sensor.home_energy_gateway_battery')|int(0) != 0 %}
           {{states('sensor.home_energy_gateway_battery')}}
        {% else %}
           {{states('sensor.filtered_powerwall_soc')}}
        {% endif %}
      unit_of_measurement: "%"

    filtered_powerwall_battery_remaining: 
      value_template: >
        {% if states('sensor.home_energy_gateway_battery_remaining')|int(0) != 0 %}
           {{states('sensor.home_energy_gateway_battery_remaining')}}
        {% else %}
           {{states('sensor.filtered_powerwall_battery_remaining')}}
        {% endif %}      
      unit_of_measurement: Wh
```

**Comment by monkga1** [comment]
Thanks I'm new to Home Assistant so haven't as yet tried anything involving actual code.  Will take a look sometime.

Regards
Gav
________________________________

**Comment by jaymunro** [comment]
Yes that would naturally filter the erroneous responses, but it does not address the matter of the random(ish) timestamps on the returned results. This randomness even suggests to me that the supposed correct returns may not actually be correct and actually refer to other times than the present time.

**Comment by Duff-Box1** [comment]
I have this issue as well using the new 3.10.3
Is there a fix for this that I can implement?

**Comment by purcell-lab** [comment]
> I have this issue as well using the new 3.10.3 Is there a fix for this that I can implement?

You can find my workaround above.

https://github.com/alandtse/tesla/issues/475#issuecomment-1406097199

**Comment by Duff-Box1** [comment]
Thanks, looks useful. I actually just reinstalled the other Tesla powerwall app as well and that battery level sensor seems ok, seems to coexist ok with the Tesla custom integration and I think has a few more sensors. Will see how this set up behaves.Thanks for the input.Sent from my iPadOn 15 Mar 2023, at 08:57, purcell-lab ***@***.***> wrote:﻿

I have this issue as well using the new 3.10.3 Is there a fix for this that I can implement?

You can find my workaround above.
#475 (comment)

**Comment by jaymunro** [comment]
I just stopped using this for battery levels and power draw and went back to using the PowerWall integration. 

The PowerWall integration is local with no internet dependency so latency is much better. 

However this integration does supply extra functionality, such as backup reserve, grid charging, operation mode which need the Tesla app on phone to adjust if not using. 

AND it provides extra functionality not even available on the Tesla app
 - Energy exports (‘Solar’ or ‘Everything’). ‘Solar’ stops the gateway exporting any battery energy to the grid.

Having both installed works well for me and provides the best of both worlds.

**Comment by Duff-Box1** [comment]
Agree this is a good solution to use both, maybe one day one integration will cover everything. It’s great work from the contributors who make this all happen.Sent from my iPadOn 16 Mar 2023, at 18:49, John ***@***.***> wrote:﻿
I just stopped using this for battery levels and power draw and went back to using the PowerWall integration.
The PowerWall integration is local with no internet dependency so latency is much better.
However this integration does supply extra functionality, such as backup reserve, grid charging, operation mode which need the Tesla app on phone to adjust if not using.
AND it provides extra functionality not even available on the Tesla app

Energy exports (‘Solar’ or ‘Everything’). ‘Solar’ stops the gateway exporting any battery energy to the grid.

Having both installed works well for me and provides the best of both worlds.

**Comment by SteveWin1** [comment]
I had the same issue and fixed it by ignoring zero results, kind of similar to purcell-lab's solution, but without using yaml templates, since I'm much more comfortable with python than yaml. I just modified the sensor.py file by replacing this line:
```python
        return round(self._energysite.percentage_charged)
```
with the following:
```python
        if self._energysite.percentage_charged > 0: 
          return round(self._energysite.percentage_charged)
        else:
          return round(float(hass.states.get('sensor.powerwalls_battery').state))
```
and replacing this line:
```python
        return round(self._energysite.energy_left)
```
with the following:
```python
        if self._energysite.energy_left > 0:
          return round(self._energysite.energy_left)
        else:
          return round(float(hass.states.get('sensor.powerwalls_battery_remaining').state))
```
That solved the issue for me and allows me to just use what comes from the integration, rather than having a template and two extra entities. Obviously change out "sensor.powerwalls_battery" and "sensor.powerwalls_battery_remaining" if your entity_id is different.

**Comment by jaymunro** [comment]
This looks to be resolved by #724
